IWSLT 2022 speech translation recipe

mzboito opened the pull request:
This is a recipe for wav2vec 2.0 fine-tuning in the speech translation task. It includes data processing for the Tamasheq-French dataset, and the parameters from the best system in the low-resource task (that will be used as baseline next year).

TParcollet commented:
Hey @mzboito did you finally solved your issues ? I won't have time to have a look at all that before a while I am afraid :/

mzboito commented:
Dear reviewers. 
I updated some headers, removed some imports and added information at the tests/recipes.csv. 
I'm unsure about why the pre-commits are failing. It's my first time doing a PR, so I'll appreciate your guidance.

When I try to run "pytest tests" locally, I get an error related to fairseq's progress_bar.
This problem seems to happen even when I'm testing the main branch of speechbrain. Do I need an specific fairseq distribution in order to run the tests? (btw, my recipe doesn't use fairseq)

Thanks for your time.

mzboito commented:
Hello again. Sorry about the mess! I fixed my formatting and now "pre-commit run --all-files" passes on my machine.

TParcollet commented on recipes/IWSLT22_lowresource/data_proc/prepare_tamasheq.sh:
```diff
@@ -0,0 +1,20 @@
+
+# 2. train the tokenizer
+# /!\ it requires the command line extension for sentence piece, available here: https://github.com/google/sentencepiece
+spm_train --input IWSLT2022_Tamasheq_data/taq_fra_clean/train/txt/train.fra  --vocab_size=1000 --model_type=unigram --model_prefix=IWSLT2022_Tamasheq_data/taq_fra_clean/train/spm_unigram1000
```
Hi ! Can't we make this compatible with SB? By this, I mean that we can already train a tokenizer with SB. It is not usual for a us to have a shell script, and in fact, we don't really need it. The whole tokenizer pre-training and data_preparation can be done within a single train.py file :-)

TParcollet commented on recipes/IWSLT22_lowresource/data_proc/to_json.py:
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python3
+"""
+Tamasheq-French data processing.
```
we will have to use the same naming convention that usual for this script i.e iwslt22_prepare.py

TParcollet commented on recipes/IWSLT22_lowresource/train.py:
```diff
@@ -0,0 +1,400 @@
+        if hparams["debug"]:
+            datasets["train"] = datasets["train"].filtered_sorted(
+                key_min_value={"duration": 1},
+                key_max_value={"duration": 5},
```
We should avoid have values hardcoded. COuld 1 and 5 be put in the yaml as hyperparams instead?

TParcollet reviewed (changes requested):
Hi @mzboito and thank you so much for this work! Once my tree comment will be fixed, we will be able to merge! Thanks again!

mzboito commented:
Hi Titouan, thanks for the corrections! I did all the changes and retrained a model to make sure nothing breaks. 
All my files pass the pre-commit with no problem however, I'm unable to commit because the pre-push is testing other files (that were not changed by me) and generating an error at a different folder. Can you help me? I have no idea how to avoid this check in order to push the last version of my code. I'm attaching an image.

The error is at speechbrain/alignment/ctc_segmentation.py

![image](https://user-images.githubusercontent.com/11521062/187646449-03aa6e4a-dc62-436f-a723-576a4b6db121.png)

mzboito commented:
Hi Titouan, thanks for the corrections! I did all the changes and retrained a model to make sure nothing breaks.
All my files pass the pre-commit with no problem however, I'm unable to commit because the pre-push is testing other files (that were not changed by me) and generating an error at a different folder. Can you help me? I have no idea how to avoid this check in order to push the last version of my code. I'm attaching an image.
I also ran a "pytests tests" without any problem (only a few warnings).

The error is at speechbrain/alignment/ctc_segmentation.py, which I do not use.

![image](https://user-images.githubusercontent.com/11521062/187650952-71025723-3d97-47ff-b946-bfe43bce7043.png)
![image](https://user-images.githubusercontent.com/11521062/187650981-662af408-ac94-4dc3-bddb-c9301a6d1510.png)

mzboito commented:
Hello Titouan, thanks for all your comments! I applied all the changes and trained a model from scratch to make sure nothing breaks with the new tokenizer integrated into the train.py.
My code passes the pre-commit and the pytest tests executes without a problem, but when I tried to commit, I got a bunch of errors about the syntax of unrelated files. After some obscure git commands, I managed to push the new version. I hope everything works!
Thanks again!

anautsch commented:
> The error is at speechbrain/alignment/ctc_segmentation.py, which I do not use.

Hi @mzboito known issue. There are scripts in the tests folder that shadow this. You can try to run locally:
```
./tests/.run-linters.sh
./tests/.run-unittests.sh
./tests/.run-doctests.sh
```

Just saw your message while I drafted this one - re-starting the tests, let's see.

TParcollet commented on recipes/IWSLT22_lowresource/README.md:
````diff
@@ -0,0 +1,57 @@
+For preparing the dataset, please run this command:
+
+```
+bash data_proc/prepare_iwslt22.sh
````
Thanks for the changes @mzboito but I will request some more because I am pretty "chiant" :p The only issue remaining is the data preparation. We do not have any shell script, hence, this part should disappear. You can put the downloading (git clone) instruction directly on the README file on a proper section (#Data access or whatever). Then, we do not prepare the data by calling an external script, but rather, directly in the main function (see bellow comment).

TParcollet commented on recipes/IWSLT22_lowresource/train.py:
```diff
@@ -0,0 +1,409 @@
+        hyperparams_to_save=hparams_file,
+        overrides=overrides,
+    )
+
```
Usually, we call the data preparation here with a proper run_on_main call (see other recipes) :-)

anautsch replied:
@mzboito can the data preparation be moved/called here via `run_on_main` ?

mzboito replied:
Hello @anautsch , can you please check if you have the last version of my code?
I already did this change: see exchange below with Titouan and line 391 here https://github.com/mzboito/speechbrain/blob/iwslt_speech_translation/recipes/IWSLT22_lowresource/train.py

anautsch replied:
Hi @mzboito quick response !
Github didn't mark the file as outdates, but you are right. I have two tabs open right now, one with the change and another one without - sorry - I'll resolve this one, you are right.

TParcollet reviewed (changes requested):
One more change, and we're good to go! Thanks @mzboito

mzboito commented:
Hi again Titouan! Thanks for the feedback: it looks much better like this. :)
I incorporated all your changes: 
1. added git clone to read me
2. removed data_proc folder
3. moved the python script to the root folder for this recipe
4. added call to this script using run_on_main at train.py
I hope everything is ok now. Let me know otherwise! :)

TParcollet commented:
Nice! Lemme try it!

mzboito commented:
Oh I'm sorry, the error is that I didn't update the recipe csv file!!

anautsch commented on recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml:
```diff
@@ -0,0 +1,197 @@
+train_log: !ref <output_folder>/train_log.txt
+
+# root data folder points to 17h version inside the github folder
+root_data_folder: IWSLT2022_Tamasheq_data/taq_fra_clean/
```
I like that the data folder hierarchy is splitted by variables !

Should a top-level data folder variable be included with `!PLACEHOLDER` ?
In other recipes, the placeholder is a reminder to users to adjust for their specific environments/paths.

anautsch reviewed (changes requested):
it's just for the minor comments. lgtm otherwise !

( obligatory curiosity question: do you plan to upload the models ? )

anautsch commented on recipes/IWSLT22_lowresource/train.py:
```diff
@@ -0,0 +1,420 @@
+        # use smaller dataset to debug the model
+        if hparams["debug"]:
+            datasets["train"] = datasets["train"].filtered_sorted(
+                key_min_value={"duration": 3},
```
Going through Titouan's earlier comments and the current version. Here's a hard-coded value that could be read from hparams.

anautsch commented on recipes/IWSLT22_lowresource/train.py:
```diff
@@ -0,0 +1,420 @@
+    )
+
+    # Data preparation
+    from recipes.IWSLT22_lowresource import prepare_iwslt22
```
Can it be loaded directly ?
e.g. `import prepare_iwslt22`

I did not test it - it's just to have it similar to other recipes as in:
https://github.com/mzboito/speechbrain/blob/6519b160d9fe1e9e20c0bd71fded36827cc73609/recipes/LibriSpeech/ASR/seq2seq/train.py#L355

anautsch commented:
GitHub indicates a conflict with the file `tests/recipes.csv` - can you please fetch the latest develop version into your PR? 
This should help resolving it. It looks like other recipes have updates (no clue though why git remarks it; it should not).

mzboito commented:
Hello @anautsch ,
Thanks for your feedback. In the last commit I did the following changes:
1. I added new recipes.csv copied from the main branch. I added my recipe at the end.
2. Replaced the value inside debug by a parameters defined at the hparams file
3. Changed the import for the data_proc as requested.

I hope it works fine now! For some reason github is still saying there's a conflict at recipes.csv

anautsch commented:
> Hello @anautsch , Thanks for your feedback. In the last commit I did the following changes:
> 
> 1. I added new recipes.csv copied from the main branch. I added my recipe at the end.
> 2. Replaced the value inside debug by a parameter defined at the hparams file
> 3. Changed the import for the data_proc as requested.
> 
> I hope it works fine now! For some reason github is still saying there's a conflict at recipes.csv

Thanks @mzboito !
It literally was just this now
```
<<<<<<< iwslt_speech_translation
recipe0144,ST,Tamasheq-French,recipes/IWSLT22_lowresource/train.py,recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml,recipes/IWSLT22_lowresource/prepare_iwslt22.py,recipes/IWSLT22_lowresource/README.md,,,,
=======
>>>>>>> develop
```
before, more lines were affected. It could be simply a tree/history versioning thingy (github still thinking of your version at branch time, and now this one being copied over - so a resolve merge cleared it up - thanks for preparing this, so it was easy!)

anautsch commented:
@mzboito linters got to be kidding...
```
Trim Trailing Whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml
```

mzboito commented:
Sorry @anautsch , I'm in a new machine and i forgot to install black and flake8!
The trailing character should be gone now.

anautsch commented:
it's about trailing whitespaces, the other linters passed.
```
Fix End of Files.........................................................Passed
Fix requirements.txt.....................................................Passed
Mixed line ending........................................................Passed
Check for added large files..............................................Passed
black....................................................................Passed
flake8...................................................................Passed
yamllint.................................................................Passed
```

mzboito commented:
That's a bit odd, I don't know why linters is mad about my comment, but I moved it somewhere else. Let's see!

anautsch commented:
> That's a bit odd, I don't know why linters is mad about my comment, but I moved it somewhere else. Let's see!

Indeed. Heading for lunch; fingers crossed !

mzboito commented:
Bon app @anautsch ! Locally, when I run black giving as input the hparams file I get the following error:
error: cannot format recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml: Cannot parse: 13:11: __set_seed: !!python/object/apply:torch.manual_seed [!ref <seed>]
All done! 💥 💔 💥
1 file failed to reformat.

I don't understand why just now it is complaining about this line. Before it was running fine. Moreover, I checked a different recipe, and the line is identical (e.g. https://github.com/speechbrain/speechbrain/blob/develop/templates/speaker_id/train.yaml)

Not sure how to fix this.

anautsch commented:
> Bon app ! 

Thanks; let's see what we have here.

> Locally, when I run black giving as input the hparams file I get the following error [...] I don't understand why just now it is complaining about this line.

black should handle py files only; cf linters: `git ls-files | grep -E "\.py$" | xargs black --check --diff`

> I checked a different recipe, and the line is identical (e.g. https://github.com/speechbrain/speechbrain/blob/develop/templates/speaker_id/train.yaml)

It has the same error when given to black.

> Not sure how to fix this.

```
yamllint recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml
```
gives

```
  19:81     warning  line too long (106 > 80 characters)  (line-length)
  31:81     warning  line too long (102 > 80 characters)  (line-length)
  42:81     warning  line too long (93 > 80 characters)  (line-length)
  75:81     warning  line too long (89 > 80 characters)  (line-length)
  198:61    error    no new line character at the end of file  (new-line-at-end-of-file)
```
When I added an empty line at the end that error disappeared (but it wasn't reported before either).

Chasing down this lead: `grep -r trailing-whitespace .` 
=> `.pre-commit-config.yaml` points to https://github.com/pre-commit/pre-commit-hooks

So, I tried out: `pre-commit run trailing-whitespace --files recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml`
```
Trim Trailing Whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml
```

ran a `meld` on the supposedly fixed one and another local copy: `Files are identical` !!
Ran it a second time:
```
pre-commit run trailing-whitespace --filesml recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml
Trim Trailing Whitespace.................................................Passed
```

Idk - some cache issue?

Please run twice:
```
pre-commit run trailing-whitespace --filesml recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml
```

anautsch commented:
This is funny and frightening at the same time ... what is git up to...
```
$ git diff
diff --git a/recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml b/recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml
index a889f2e9..63fb2b10 100644
--- a/recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml
+++ b/recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml
@@ -42,7 +42,7 @@ ckpt_interval_minutes: 15 # save checkpoint every N min
 # Data sorting parameters: sorting_debug_duration replaces sorting_min_duration in debug mode
 sorting: ascending
 sorting_min_duration: 1
-sorting_debug_duration: 3 
+sorting_debug_duration: 3
 sorting_max_duration: 5

```

mzboito commented:
Hello @anautsch , sorry for the delay.
I think the problem was an extra space after the number 3. Sorry about that.
Now it's passing pre-commit.

pre-commit run trailing-whitespace --files recipes/IWSLT22_lowresource/hparams/train_w2v2_st.yaml
Trim Trailing Whitespace.................................................Passed

anautsch commented on recipes/IWSLT22_lowresource/README.md:
````diff
@@ -0,0 +1,56 @@
+
+For launching training:
+```
+python train.py hparams/w2v2_st.yaml
````
@mzboito the yaml file name changed (sorry I spot it only now) -> hparams/train_w2v2_st.yaml

mzboito replied:
Nice catch! I fixed it in my last commit. Thanks! :)

anautsch reviewed (changes requested):
one last bit - just went through a final time
(never ending story here... sorry for that)

_(about the white space, at some point, I need to reconfigure my tools and/or clean my glasses... expected it but didn't catch it before)_
